crates/sol-macro-gen/src/sol_macro_gen.rs

```diff
@@ -0,0 +1,367 @@
+//! SolMacroGen and MultiSolMacroGen
+//!
+//! This type encapsulates the logic for expansion of a Rust TokenStream from Solidity tokens. It
+//! uses the `expand` method from `alloy_sol_macro_expander` underneath.
+//!
+//! It holds info such as `path` to the ABI file, `name` of the file and the rust binding being
+//! generated, and lastly the `expansion` itself, i.e the Rust binding for the provided ABI.
+//!
+//! It contains methods to read the json abi, generate rust bindings from the abi and ultimately
+//! write the bindings to a crate or modules.
+
+use alloy_json_abi::JsonAbi;
+use alloy_sol_macro_expander::expand::expand;
+use alloy_sol_macro_input::{tokens_for_sol, SolInput, SolInputKind};
+use eyre::{Ok, OptionExt, Result};
+use foundry_common::fs;
+use proc_macro2::{Ident, Span, TokenStream};
+use serde_json::Value;
+use std::{
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+pub struct SolMacroGen {
+    pub path: PathBuf,
+    pub name: String,
+    pub expansion: Option<TokenStream>,
+}
+
+impl SolMacroGen {
+    pub fn new(path: PathBuf, name: String) -> Self {
+        Self { path, name, expansion: None }
+    }
+
+    pub fn get_json_abi(&self) -> Result<(JsonAbi, Option<String>)> {
+        let json = std::fs::read(&self.path)?;
+
+        // Need to do this to get the abi in the next step.
+        let json: Value = serde_json::from_slice(&json)?;
+
+        let abi_val = json.get("abi").ok_or_eyre("No ABI found in JSON file")?;
+        let json_abi = serde_json::from_str(&abi_val.clone().to_string())?;
+
+        let bytecode = json.get("bytecode").and_then(|b| b.get("object")).map(|o| o.to_string());
```

Comment:
`JsonAbi` can be deserialized directly. Please see the logic in sol-macro-input

in fact I would prefer if you just passed the path through `quote!` to syn and parse a SolInput directly